numpydoc/xref.py

```diff
@@ -0,0 +1,169 @@
+    #  - dict(str, int)
+    #  - dict[str, int]'
+    #  - tuple(float, float)
+    #  - dict[tuple(str, str), int]'
+    r'^'
+    r'(dict|list|tuple)'
+    r'[\[\(]'
+    r'(.+?(?:,\s*)?)+'
+    r'[\]\)]'
+    r'$'
+)
+
+CONTAINER_SPLIT_RE = re.compile(
+    # splits dict(str, int) into
+    #    ['dict', '[', 'str', ', ', 'int', ']', '']
+    r'(\s*[\[\]\(\),]\s*)'
+)
+
+DOUBLE_QUOTE_SPLIT_RE = re.compile(
+    # splits 'callable ``f(x0, *args)`` or ``f(x0, y0, *args)``' into
+    #    ['callable ', '``f(x0, *args)``', ' or ', '``f(x0, y0, *args)``', '']
+    r'(``.+?``)'
+)
+
+IGNORE = {'of', ' of ', 'either', 'or', 'with', 'in', 'default'}
+CONTAINER_CHARS = set('[](){}')
+
+
+def make_xref_param_type(param_type, xref_aliases):
+    """
+    Enclose str in a role that creates a cross-reference
+
+    The role ``xref_param_type`` *may be* added to any token
+    that looks like type information and no other. The
+    function tries to be clever and catch type information
+    in different disguises.
+
+    Parameters
+    ----------
+    param_type : str
+        text
+    xref_aliases : dict
+        Mapping used to resolve common abbreviations and aliases
+        to fully qualified names that can be cross-referenced.
+
+    Returns
+    -------
+    out : str
+        Text with parts that may be wrapped in a
+        ``xref_param_type`` role.
+    """
+    if param_type in xref_aliases:
+        param_type = xref_aliases[param_type]
+
+    if (QUALIFIED_NAME_RE.match(param_type) and
+            param_type not in IGNORE):
+        return ':xref_param_type:`%s`' % param_type
+
+    # Clever stuff below (except the last return)
+    # can be removed without affecting the basic functionality.
+
+    def _split_and_apply_re(s, pattern):
+        """
+        Split string using the regex pattern,
+        apply main function to the parts that do not match the pattern,
+        combine the results
+        """
+        results = []
+        tokens = pattern.split(s)
+        if len(tokens) > 1:
+            for tok in tokens:
+                if pattern.match(tok):
+                    results.append(tok)
+                else:
+                    results.append(
+                        make_xref_param_type(tok, xref_aliases))
+
+            return ''.join(results)
+        return s
+
+    def _split_and_apply_str(s, on):
+        """
+        Split string s, at the substring on,
+        apply main function to the splits,
+        combine the results
+        """
+        return on.join(
+            make_xref_param_type(s, xref_aliases)
+            for s in s.split(on))
+
+    # The cases are dealt with in an order the prevents
+    # conflict.
+    # Then the strategy is:
+    #   - Identify a pattern we are not interested in
+    #   - split off the pattern
+    #   - re-apply the function to the other parts
+    #   - join the results with the pattern
+
+    # endswith ', optional'
+    if param_type.endswith(', optional'):
```

Comment:
Does this strictly need to be at the end? is there a reason not to just include `optional` in ignore?